File: src/python/jeans_initial.py
import numpy as np
import matplotlib.pyplot as plt
from scipy.stats import norm

# Plummer profile
def plummer_density_profile(r, a, M=1.0):
    return (3 * M) / (4 * np.pi * a**3) * (1 + (r / a)**2)**(-2.5)

def dphi_dr_plummer(r, a,G=1.0, M=1.0):
    return G * M * r / np.sqrt(r**2 + a**2)**3.0

def phi_plummer(r, a,G=1.0, M=1.0):
    return -G*M/np.sqrt(r**2+a**2)

def plummer_velocity_dispersion(r,a,M=1.0,G=1.0):
    return np.sqrt(G * M / np.sqrt(r**2 + a**2)/6.0)

def create_particles_single(N_particles, box_size, a, M, mode, solver,G=1.0):
    boundary = solver
    # --- Step 1: Generate positions using Plummer distribution
    box_center = np.array([box_size / 2] * 3)
    u = np.random.rand(N_particles)
    r = a * (u**(-2/3) - 1)**(-0.5)  # radial distribution from inverse CDF

    theta = np.arccos(1 - 2 * np.random.rand(N_particles))
    phi = 2 * np.pi * np.random.rand(N_particles)

    x = r * np.sin(theta) * np.cos(phi)
    y = r * np.sin(theta) * np.sin(phi)
    z = r * np.cos(theta)
    positions = np.stack([x, y, z], axis=1) + box_center

    # --- Step 2: Compute velocity dispersion at each radius
    if mode == "stable":
        scale = 1.0
    elif mode == "contract":
        scale = 0.5
    elif mode == "expand":
        scale = 1.5
    else:
        raise ValueError(f"Unknown mode '{mode}'")

    sigma = plummer_velocity_dispersion(r, a, M, G)
    velocities = np.random.normal(0, 1, size=(N_particles, 3)) * (scale * sigma[:, np.newaxis])

    # --- Step 3: Assign equal mass to all particles
    masses = np.full(N_particles, M / N_particles)

    if boundary == 'isoalated':
        mask = np.all((positions >= 0) & (positions < box_size), axis=1)
        positions = positions[mask]
        velocities = velocities[mask]
        masses = masses[mask]

    return positions, velocities, masses

def create_particles_double(N_particles, box_size, a, M, mode, solver,G, add_initial_velocity , v_offset):
    N_each = N_particles // 2
    positions_all = []
    velocities_all = []
    boundary = solver

    for center_shift, bulk_v in [(-0.1, +v_offset), (+0.1, -v_offset)]:
        box_center = np.array([box_size / 2 + center_shift, box_size/2, box_size/2])

        u = np.random.rand(N_each)
        r = a * (u**(-2/3) - 1)**(-0.5)

        theta = np.arccos(1 - 2 * np.random.rand(N_each))
        phi = 2 * np.pi * np.random.rand(N_each)

        x = r * np.sin(theta) * np.cos(phi)
        y = r * np.sin(theta) * np.sin(phi)
        z = r * np.cos(theta)
        positions = np.stack([x, y, z], axis=1) + box_center

        # velocity dispersion
        if mode == "stable":
            scale = 1.0
        elif mode == "contract":
            scale = 0.5
        elif mode == "expand":
            scale = 1.5
        else:
            raise ValueError(f"Unknown mode '{mode}'")

        sigma = plummer_velocity_dispersion(r, a, M, G)
        velocities = np.random.normal(0, 1, size=(N_each, 3)) * (scale * sigma[:, np.newaxis])

        
        if add_initial_velocity:
            velocities[:, 0] += bulk_v 

        positions_all.append(positions)
        velocities_all.append(velocities)

    positions_all = np.vstack(positions_all)
    velocities_all = np.vstack(velocities_all)
    masses = np.full(N_particles, M / N_particles)

    if boundary == 'isoalated':
        mask = np.all((positions >= 0) & (positions < box_size), axis=1)
        positions = positions[mask]
        velocities = velocities[mask]
        masses = masses[mask]

    return positions_all, velocities_all, masses

File: src/C/jeans_initial.cpp
#include "jeans_initial.hpp"
#include <random>
#include <cmath>
#include <vector>
#include <array>
#include <string>
#include <stdexcept>
#include <iostream>
#include <ostream>
#include <algorithm>

using ParticleArray = std::vector<std::array<double, 3>>;
using MassArray = std::vector<double>;

namespace {
    double plummer_velocity_dispersion(double r, double a, double M, double G = 1.0) {
        return std::sqrt(G * M / std::sqrt(r*r + a*a) / 6.0);
    }
}

ParticleArray create_plummer_positions(int N, double box_size, double a, double shift) {
    ParticleArray positions(N);
    std::mt19937 gen(std::random_device{}());
    std::uniform_real_distribution<double> uni(0.0, 1.0);

    for (int i = 0; i < N; ++i) {
        double u = uni(gen);
        u = std::clamp(u, 1e-6, 1.0 - 1e-6);  // avoid u = 0 or 1

        double temp = std::pow(u, -2.0/3.0) - 1.0;
        double r = a / std::sqrt(temp);     // Plummer radius


        // Sample direction uniformly on the sphere
        double theta = std::acos(1 - 2 * uni(gen));
        double phi = 2 * M_PI * uni(gen);

        // Now convert (r, θ, φ) to Cartesian if needed
        double x = r * std::sin(theta) * std::cos(phi);
        double y = r * std::sin(theta) * std::sin(phi);
        double z = r * std::cos(theta);

        positions[i] = {x + box_size/2 + shift, y + box_size/2, z + box_size/2};
    }
    return positions;
}

ParticleArray create_plummer_velocities(
    const std::vector<double>& r, double a, double M, const std::string& mode, bool add_bulk, double bulk_v)
{
    int N = r.size();
    ParticleArray velocities(N);
    std::mt19937 gen(std::random_device{}());
    std::normal_distribution<double> normal(0.0, 1.0);

    double scale = (mode == "stable" ? 1.0 : (mode == "contract" ? 0.5 : (mode == "expand" ? 1.5 : -1.0)));
    if (scale < 0.0) throw std::invalid_argument("Unknown mode: " + mode);

    for (int i = 0; i < N; ++i) {
        double sigma = plummer_velocity_dispersion(r[i], a, M);
        for (int d = 0; d < 3; ++d) {
            velocities[i][d] = normal(gen) * sigma * scale;
        }
        if (add_bulk) velocities[i][0] += bulk_v;
    }
    return velocities;
}

std::tuple<ParticleArray, ParticleArray, MassArray> create_particles_single(
    int N_particles, double box_size, double a, double M,
    const std::string& mode, const std::string& solver, double G)
{
    ParticleArray pos = create_plummer_positions(N_particles, box_size, a, 0.0);

    std::vector<double> radii(N_particles);
    for (int i = 0; i < N_particles; ++i) {
        auto& p = pos[i];
        double dx = p[0] - box_size/2;
        double dy = p[1] - box_size/2;
        double dz = p[2] - box_size/2;
        radii[i] = std::sqrt(dx*dx + dy*dy + dz*dz);
    }
    ParticleArray vel = create_plummer_velocities(radii, a, M, mode, false, 0.0);
    MassArray mass(N_particles, M / N_particles);

    if (solver == "isolated") {
        ParticleArray p_new, v_new;
        MassArray m_new;
        for (int i = 0; i < N_particles; ++i) {
            const auto& p = pos[i];
            if ((p[0]>=0 && p[0]<box_size) && (p[1]>=0 && p[1]<box_size) && (p[2]>=0 && p[2]<box_size)) {
                p_new.push_back(p);
                v_new.push_back(vel[i]);
                m_new.push_back(mass[i]);
            }
        }
        return {p_new, v_new, m_new};
    }
    return {pos, vel, mass};
}

std::tuple<ParticleArray, ParticleArray, MassArray> create_particles_double(
    int N_particles, double box_size, double a, double M,
    const std::string& mode, const std::string& solver, double G,
    bool add_initial_velocity, double v_offset)
{
    int N_each = N_particles / 2;
    ParticleArray pos_all, vel_all;
    MassArray mass(N_particles, M / N_particles);

    for (auto [shift, bulk_v] : std::vector<std::pair<double, double>>{{-0.1, +v_offset}, {+0.1, -v_offset}}) {
        ParticleArray pos = create_plummer_positions(N_each, box_size, a, shift);

        std::vector<double> radii(N_each);
        for (int i = 0; i < N_each; ++i) {
            auto& p = pos[i];
            double dx = p[0] - box_size/2;
            double dy = p[1] - box_size/2;
            double dz = p[2] - box_size/2;
            radii[i] = std::sqrt(dx*dx + dy*dy + dz*dz);
        }
        ParticleArray vel = create_plummer_velocities(radii, a, M, mode, add_initial_velocity, bulk_v);
        pos_all.insert(pos_all.end(), pos.begin(), pos.end());
        vel_all.insert(vel_all.end(), vel.begin(), vel.end());
    }

    if (solver == "isolated") {
        ParticleArray p_new, v_new;
        MassArray m_new;
        for (size_t i = 0; i < pos_all.size(); ++i) {
            const auto& p = pos_all[i];
            if ((p[0]>=0 && p[0]<box_size) && (p[1]>=0 && p[1]<box_size) && (p[2]>=0 && p[2]<box_size)) {
                p_new.push_back(p);
                v_new.push_back(vel_all[i]);
                m_new.push_back(mass[i]);
            }
        }
        return {p_new, v_new, m_new};
    }
    return {pos_all, vel_all, mass};
}